q2_cutadapt/_demux.py

```diff
@@ -96,41 +98,49 @@ def _write_empty_fastq_to_mux_barcode_in_seq_fmt(seqs_dir_fmt):
         seqs_dir_fmt.file.write_data(fastq, FastqGzFormat)
 
 
-def _demux(seqs, barcodes, error_tolerance, untrimmed):
+def _demux(seqs, barcodes, error_tolerance, mux_fmt, batch_size):
     barcodes = barcodes.to_series()
     per_sample_sequences = CasavaOneEightSingleLanePerSampleDirFmt()
-
-    _write_empty_fastq_to_mux_barcode_in_seq_fmt(untrimmed)
-
-    with tempfile.NamedTemporaryFile() as barcode_fasta:
-        _write_barcode_fasta(barcodes, barcode_fasta)
-        cmd = _build_demux_command(seqs, barcode_fasta, per_sample_sequences,
-                                   untrimmed, error_tolerance)
-        run_command(cmd)
-
+    n_samples = len(barcodes)
+    if batch_size > n_samples:
+        raise ValueError('batch_size needs to be less than or equal to the '
```

Comment:
```suggestion
        raise ValueError('batch_size must be less than or equal to the '
```

Reply:
How about something like this?

```python
raise ValueError('The batch_size (%d) cannot be greater than the number of samples (%d).')
```

---

I wish we could dynamically set the `Range` on the method, that way we don't need this guard. Another alternative is to do a way with the guard altogether and just clamp the batch_size to the number of samples. I don't like that though, because it isn't clear to the user that that is what is happening (as well, provenance wouldn't record that, either).

Reply:
Love it.